### ISSUE TITLE
Parallelize the deployment of hubs on a cluster

### ISSUE THREAD

**Issue by choldgraf** [2i2c-org/infrastructure#818]
### Description

We should deploy our hubs in parallel during CI/CD. This would make the deployment of each hub faster and remove unnecessary interdependence between hubs.

### Problem and value in solving it

Deploying in serial has a few shortcomings:

- Hubs at the end of the queue must wait for previous hubs to finish deploying before they are deployed. This is a problem if we really need a hub at the end of a queue to be deployed quickly (like during an event)
- If a hub's deployment fails (e.g., because of #586 ) then all subsequent hubs will *not* be deployed at all, even if they should be.

Parallelizing our hub deploys should make the whole thing faster, and also removes the dependence of one hub's deployment on all of the other hub deployments.

### Design constraints

- **Hubs should deploy in parallel**.
- **Failure for one hub should not stop deployment on other hubs**.
- **A hub should only deploy if configuration relevant to it has been changed**.
- **Use staging hub tests to make sure we are ready to deploy to prod**. While we want to parallelize the deployment of community hubs, we should continue deploying *staging hubs* first and intentionally block deployment for a given hub if the staging hub fails to pass its tests.

### Implementation ideas

- We could investigate chained GitHub actions jobs to define the dependence between a staging deploy + tests, and a subsequent deploy to prod at the hub-level

### Related issues

- https://github.com/2i2c-org/infrastructure/issues/586
- https://github.com/2i2c-org/infrastructure/issues/857
- https://github.com/2i2c-org/infrastructure/issues/279

**Comment by damianavila** [comment]
One thing we have been discussing on Slack that is relevant here is the idea of "maybe selectively deploying hubs only if certain config changes"

@sgibson91 replied
> I think this would be really hard to achieve as it would likely involve reading the YAML in the workflow

I agree with that idea unless we make changes to be granular...

> Wondering if more granularity would help us here, I mean... instead of having a big config file for the cluster, let's have a directory to represent the cluster plus 1) a cluster-specific yaml file and 2) several hub-specific yaml files for each staging/prod pair.

> See the pre-existing discussion we had here: https://github.com/2i2c-org/infrastructure/issues/124

> With more granularity, it would be definitely easier to come up with a proper matrix that selectively deploys on each hub pair accordingly to the specific config we are touching... in addition, it actually helps with our R2R story because the config is actually "isolated" from others if they want to migrate.

And @choldgraf replied

> another benefit would be that you could create a permalink to somebody's hub configuration

With more granularity, we should also have improvements on the CI testing side, ref: https://github.com/2i2c-org/infrastructure/issues/797#issuecomment-957299658 and https://github.com/2i2c-org/infrastructure/issues/586

**Comment by choldgraf** [comment]
Just to clarify that in https://github.com/2i2c-org/infrastructure/issues/124 there was general agreement that splitting the YAML configuration for each hub was a good idea 👍 

Shall we scope that inside this issue as well? I could see how we'd benefit from doing both at the same time.

**Comment by sgibson91** [comment]
> Shall we scope that inside this issue as well? I could see how we'd benefit from doing both at the same time.

Good idea!

**Comment by consideRatio** [comment]
I arrive here from #857 and have some ideas of what I think is important for any chosen structure.

- Have helm chart config files in standalone file, this means one per hub installation
- Don't have deployer script distinguish prod and staging hubs in a hardcoded manner
  I'm not sure if this is done, but if it is I think we should stop and instead introduce more general concepts like hubs that depend on other hubs
- Use a folder per k8s cluster
  It will have a lot of related config that needs to be in separate fles, so it would probably be easier to navigate if its nested inside a folder.
- look to be explicit rather than abbreviated in file/folder names
  As an example, this is very explicit.
  ```
  |- config/
    |- {{ cluster_name.k8s-cluster }}/
      |- <hub-name>.hub.deployer-config.yaml
      |- <hub-name>.hub.helm-chart-config.yaml
      |- <k8s-cluster-name>.k8s-cluster.deployer-config.yaml
  ```
  With explicitly, I'm able to think more clearly and I think everyone will understand things more quickly. It also comes with a benefit that it is easier to enable primitive pattern matches useful in CI logic checking if certain kinds of files has changed (`**/*.helm-chart-config.yaml`, or `**/<some-name>.k8s-cluster/**`).

**Comment by yuvipanda** [comment]
> * If a hub's deployment fails (e.g., because of #586 ) then all subsequent hubs will *not* be deployed at all, even if they should be.

Fwiw, I think of this as a *feature*. We don't treat staging hubs separately except they are the *first* ones to be deployed. So if a deployment fails there for *whatever reason*, it blocks deployment to prod hubs. This helps put us in a situation where only the staging hub (per cluster) is down from a bad deploy, rather than every hub we have.

This is the closest we have to a [canary deployment](https://octopus.com/docs/deployments/patterns/canary-deployments). However we choose to parallelize the deploys, we should make sure that we have some way of defending against bad config after deployment as well in a way that minimizes blast radius.

**Comment by consideRatio** [comment]
> > * If a hub's deployment fails (e.g., because of [Improve the hub health check in deployer #586](https://github.com/2i2c-org/infrastructure/issues/586) ) then all subsequent hubs will _not_ be deployed at all, even if they should be.
> 
> Fwiw, I think of this as a _feature_. We don't treat staging hubs separately except they are the _first_ ones to be deployed. So if a deployment fails there for _whatever reason_, it blocks deployment to prod hubs. This helps put us in a situation where only the staging hub (per cluster) is down from a bad deploy, rather than every hub we have.
> 
> This is the closest we have to a [canary deployment](https://octopus.com/docs/deployments/patterns/canary-deployments). However we choose to parallelize the deploys, we should make sure that we have some way of defending against bad config after deployment as well in a way that minimizes blast radius.

I agree this is a feature but I think this can be retained in a smart way while still keeping things parallelized smartly. I'm thinking of an implementation like GitHub workflows syntax allows parallelizations of jobs: that you can describe that one job `needs` another job, and that you can have multiple workflow files separately from each other, etc.

I've thought a bit about config structure, here are wishes I'd have on the chosen structure to enable parallelization and ease of working with it:
- structure to have helm chart config in standalone files
- structure to enable re-deploy logic via git history
  - given a list of changed files, do we need to re-test / re-deploy a given helm chart?
  - note, a helm chart installation depends on: explicitly listed values files, a specific helm chart, (something else?)

**Comment by sgibson91** [comment]
Agree @yuvipanda but right now the most common reason for failure is flaky hub checks, or a hub community hasn't updated their image that causes failures (like catalyst cooperative). Other communities definitely shouldn't be penalised for that.

I think another layer of defence we can put in place is deploying to staging hubs from PRs using test-this-pr.

**Comment by damianavila** [comment]
> I think another layer of defence we can put in place is deploying to staging hubs from PRs using test-this-pr.

This is key... we are not actually testing the deployment before the merge happens...
I will totally embrace a test-this-pr workflow **effectively** deploying into `staging` before the merge happens.
I also like the idea of chained jobs, deploying first to staging and then to production but I think these two ideas could go actually in the same direction because they are not mutually exclusive.

**Comment by yuvipanda** [comment]
Yeah totally - the current 'canary' is terrible :D I just wanted to make sure that whatever we use instead we keep some form of canary.

**Comment by sgibson91** [comment]
@damianavila I have deployed both these strategies over in mybinder.org-deploy and indeed they are not mutually exclusive. And in fact, the test-this-pr workflow actually takes advantage of the fact the deploys are chained for staging and prod. Really all test-this-pr does is copy a branch from a fork into a branch of the upstream repo, so there's a lot of flexibility in terms of what we could trigger from that action.

**Comment by damianavila** [comment]
> indeed they are not mutually exclusive

Great! Thanks for the confimation!

> so there's a lot of flexibility in terms of what we could trigger from that action.

That's super interesting and eager to know what we can build on this specific topic.

**Comment by sgibson91** [comment]
Here's a question I just thought of: When parallelising deployment of _hubs_, shall we keep that just in CI? i.e. running `python deployer deploy CLUSTER_NAME` still deploys all hubs on that cluster sequentially? I think this will be important to consider when thinking about the changes to be made in the deployer when resolving #857

**Comment by consideRatio** [comment]
I'm not sure I understand the question about `shall we keep that just in CI?` but...

I think:
1. The deployer script should not attempt to deploy multiple hubs in parallel, that should only be done by running multiple deployer scripts separately in parallel (using several GitHub CI jobs for example).
2. I'm fine with the idea of deprecating the deployer scripts api to deploy an entire cluster. The value of doing so in my mind is to reduce the complexity and speed up development of the new solution. Also, for situations where a lot of hubs are deployed, we should probably rely on multiple jobs running in parallel to deploy a single hub anyhow.
3. I currently see us needing to specify an explicit dependency structure for all individual deployments in the hub, so we shouldn't need the deployment script to figure out such dependencies - but instead just assume what we have explicitly declared.

**Comment by sgibson91** [comment]
> we should probably rely on multiple jobs running in parallel to deploy a single hub anyhow.

Yes. With one issue. GitHub Actions currently doesn't support nested matrix builds 😢 so a config that looked like the below would not be possible

```yaml
strategy:
  matrix:
    include:
      - cluster_name: cluster1
         provider: gcp
         hub: ["hub1", "hub2", ... "hubN"]
```

But I have discovered that you can write out each combination by hand and have a minimum working example:

- workflow file: https://github.com/sgibson91/testing-gh-actions/blob/main/.github/workflows/includes-matrix-with-list.yaml
- output: https://github.com/sgibson91/testing-gh-actions/actions/runs/1654861757

**Comment by consideRatio** [comment]
@sgibson91 one can also let the matrix be set as a value of an output determined programmatically in a previous job! So, a bit of hazzle, but everything is doable in the end!

**Comment by sgibson91** [comment]
I think this is going to be really hard due to a multitude of limitations in the GitHub Actions framework (these may or may not also be problems across other CI vendors... but let's assume we are sticking with GitHub Actions for the time being)

- You cannot nest matrix jobs, instead you have to explicitly list each combination of parameters for each job. (Could use an installation graph generated by the deployer to do this automatically...)
- You cannot dynamically set job IDs at run time
- All sub-jobs in a matrix job gets the same job ID. This means you cannot say "I want job B to run to run with these parameters when _this specific matrix sub-job_ successfully completes".

So I think this means we're looking at something like the following workflow:

- Job 1: Install/upgrade the support chart for the cluster
- Job 2 [depends on successfully completion of Job 1]: **If** a staging hub is installed on the cluster, upgrade the release for the staging hub
- Job 3 [always dependent on Job 1, only dependent on Job 2 if staging has been detected]: For each hub, upgrade the hub's helm release

And then somehow parallelise that across clusters.

This is not optimised for communities that have more than one hub running on the same _shared_ cluster - though I don't think that's something we have yet and I'd rather push back developing around that scenario until we know it's coming down the pipeline, since this is already going to be complicated enough as it is.

Another option would be to have another mechanism handle all of this, such as a Makefile or the deployer itself, and just have a very short CI workflow file that is:

```
- install dependencies
- run tool
```

That's not nice for reasons too.

I'll stop thinking about this for a while before my brain explodes 🤯

**Comment by consideRatio** [comment]
Yeah, we have a really complicated need: we want to dynamically execute some things in sequence, and some things in parallel.

The only solution I see given the constraints are that 
1. (I strongly dislike this idea) Have an humongous workflow file with job like 100 job definitions used as placeholders for various dependency needs.
2. (I reject this idea) Have all jobs run in parallel and stall until they are ready to run somehow.
3. (I strongly dislike this idea) Have the only thing our GitHub workflow do be to trigger a non-GitHub workflow in some system that can take a set of jobs to be executed in the graph we'd like.
4. (I dislike this idea quite a bit) Have a main workflow that acts as a scheduler of other workflows it starts.
   - UX wise this means you won't be able to use GitHub's UI to see the jobs running as part of a workflow, but instead you will see many workflows with a few jobs

**Comment by GeorgianaElena** [comment]
> The tweak is to not have job 2 and 3 be separate, but a single job. The reason for it is that otherwise any delay/failure on a staging hub would delay/halt the deployments of the prod hubs. So, something like...

Maybe not deploying on the prod hub when staging fails is not a bad thing? This way only the staging hub can potentially get broken in case of an issue and the prod continues to be functional. I might be oversimplificating things though.

**Comment by sgibson91** [comment]
> hub1-staging -> hub1-prod

I think we don't really have this pattern (yet) on a shared cluster, only on dedicated clusters. I think we currently have one staging hub per cluster, with the exception of the 2i2c pilot cluster which has a basehub staging and a daskhub staging. But very few, if any, of the prod hubs on that cluster _depend_ on those staging hubs.

**Comment by damianavila** [comment]
> This is not optimised for communities that have more than one hub running on the same shared cluster - though I don't think that's something we have yet and I'd rather push back developing around that scenario until we definitely know it's coming down the pipeline, since this is already going to be complicated enough as it is.

I do see more of these shared clusters down the pipeline...

> (I dislike this idea quite a bit) Have a main workflow that acts as a scheduler of other workflows it starts.

From the top of my head, the above idea should be the guidance (regardless of the specific implementation).

We need to attack two different cases:
  1. Shared cluster with dozen of hubs
  2. Specific cluster with a pair of hubs
 
Maybe there is a way to configure actions/workflow/jobs to optimize for case 1 (deploying all the hubs in parallel) or, alternatively, configure the tools to optimize for case 2 (sequential deploy, first stating, then prod) with some config value in the cluster specification? (I might be missing additional context, so feel free to tell me: "go above and re-read the whole thing" 😛).

**Comment by sgibson91** [comment]
@damianavila the only difference I see between these two cases is the number of parallel deployments of a hub we will spawn. For a shared cluster, we parallelise over N hubs; and for a dedicated cluster, we "parallelise" over 1 hub. Each cluster (whether shared or dedicated) still has at least one staging hub, so we can still deploy in sequence there (which I believe is good for "canary" deployments). Other than that, I don't see a need to treat these particularly differently.

**Comment by sgibson91** [comment]
> I do see more of these shared clusters down the pipeline...

I'd rather work on solving the problem we have now, and then iterating for this specific example when we know it's going to happen. We do have some control here. We could say, "if you want your own staging deployment, then that requires a dedicated cluster. Otherwise, you share a general staging hub with the other hubs on a shared cluster" to make things simpler. We should discuss that more as a team if that's something we should offer or not. But I don't want that decision to block this work. I'd rather solve for what we have now in a way that is _extensible_ to work towards that scenario.

**Comment by consideRatio** [comment]
I think any decision on how we want to name values files can be tabled and followed up over time because of what we currently aim for: to explicitly declare the helm chart values files used for each release (a _release_ is Helm terminology for a helm chart installation).

I'm not very opinionated about the naming conventions used of the helm chart values files we reference, but I think it would be suitable to let it evolve over time as its not critical to get it to feel right directly.

**Comment by damianavila** [comment]
> Other than that, I don't see a need to treat these particularly differently.

I might be misunderstanding the whole thing but for me, those use cases are effectively similar at the time to technically approach them but effectively different in the way we want the deployments to happen.
In my mind, the shared cluster with N hubs is somehow deploying the N hubs in parallel if we touch the config for all those N hubs, but if we only touch hub X config, we only deploy that one without waiting for all the other ones in the same cluster to be deployed as well (except for staging, I guess, that we may want to also deploy). 

Maybe that is not possible because of the GH actions limitations you mentioned above, and if that is the case, feel free to ignore my comment.

> I'd rather solve for what we have now in a way that is extensible to work towards that scenario.

+1 on the extensibility...

**Comment by choldgraf** [comment]
> We could say, "if you want your own staging deployment, then that requires a dedicated cluster. Otherwise, you share a general staging hub with the other hubs on a shared cluster" to make things simpler.

Just a note that I think this sounds like a good idea to me

**Comment by sgibson91** [comment]
> the shared cluster with N hubs is somehow deploying the N hubs in parallel if we touch the config for all those N hubs, but if we only touch hub X config, we only deploy that one without waiting for all the other ones in the same cluster to be deployed as well

Ah, I now see we are talking about different things!

We can achieve the "only update a hub that has been changed" kind of deployed by parsing a list of changed files from GitHub Actions into a _new_ function in the deployer that can work out which hubs on which cluster should be updated. It makes sense to have this logic in the deployer since it will have the knowledge of which files are associated with a specific hub's helm installation. The output of this function will be a series of job definitions for GitHub Actions to execute. Again, it doesn't matter if this for a _shared_ cluster or a _dedicated_ cluster as the deployer has all the knowledge to work this out and will generate the jobs GHA needs to run.

---

The limitations I was speaking about with GitHub Actions are specifically to do with a (at the moment, fictional) scenario where we have a "hub pairing" on a _shared_ cluster.

So for example, we have on the same cluster:
- hub1-staging -> hub1-prod
- hub2-staging -> hub2-prod

In an _ideal_ scenario, we would like the following:

1. A matrix job that upgrades all staging releases on a cluster, and another matrix job that upgrades all production releases on a cluster
2. When the upgrade of hubN-staging successfully completes, immediately begin the upgrade process for the hubN-prod release

We **cannot** achieve 2 since all sub-jobs in a matrix job receive the same job ID, so you cannot identify when a _specifc_ sub-job has completed. So moving forwards, we have 2 options:

1. Wait for all staging releases on a cluster to successfully upgrade before beginning the production upgrades (if we stick to a one-staging-hub-per-cluster policy, I think this is fine)
2. Or @consideRatio's suggestion, where we bundle the staging upgrade and production upgrade into the same job.

---

I also think continuing to talk about this is getting complex as there are so many factors to think about. I think I should just get on with creating Pull Requests and then we will at least have an _example_ to reason about rather than everything being arbitrary conjecture.

**Comment by consideRatio** [comment]
@sgibson91 wiee nice summary!! I find this is very complicated to think about, and I think your summary above captured a lot of the complexity very nicely! :tada:

> So moving forwards, we have 2 options:

I think there are at least one more option while still relying on GitHub workflows, an option we can refer to as "dynamic workflow dispatching".

For example, it could be:
1. First do k8s cluster central updates (support chart updates) if needed
2. Second dispatch pre-defined workflows to staging upgrade (one job) -> prod upgrade (another job) for all the hubs that needs to be updated based on observed changes.

**Comment by sgibson91** [comment]
> For example, it could be:
> 
> 1. First do k8s cluster central updates (support chart updates) if needed
> 2. Second dispatch pre-defined workflows to staging upgrade (one job) -> prod upgrade (another job) for all the hubs that needs to be updated based on observed changes.

Sorry, this is what I am thinking for option one, I just dropped the ""support" bit because it wasn't pertinent to what we were talking about.

Again, I should just open a PR :sweat_smile:

**Comment by consideRatio** [comment]
> Wait for all staging releases on a cluster to successfully upgrade before beginning the production upgrades

Oh hmmmm but note how you say "wait for all staging releases". To me, that represents having a single workflow that has 3 jobs:

1. support
2. all staging (blocked by support)
3. all prod (blocked by all staging)

For me, what I suggested would be two jobs:

1. support
2. Dispatch in parallel the execution of workflows that independently and in parallel for each hub do a staging->prod release, where each hub deployment would wait for its staging part, but not for other hub deployments staging parts.

**Comment by sgibson91** [comment]
Yes, but at the minute we only have one staging hub per cluster - so that's future work

**Comment by consideRatio** [comment]
I'm very confused now, because I'm in agreement that atm we only have staging+prod. I'm thinking that PRs to do at this point are to be focused on what was outlined in the [hackmd document](https://hackmd.io/@29EjZnuAQx-ME7Olgi2JXw/Sk5S7xP6F). These are changes, that go towards enabling us to decide what specific hub deployments needs to be upgraded, which would be relevant no matter if we go with option 1, 2, or 3.

Following that, we can do implementations towards option 1, 2, or 3 as discussed in this thread.

---

Are you speaking about making a change without updating the cluster configuration, going:
- from: Support and then all hub releases are deployed in ordered sequence
- to: Support and then all hub staging releases followed by all prod releases

**Comment by sgibson91** [comment]
I am also confused. Which I why I tried to suggest we stop talking about it until there is a PR example we can reason over, because I'm finding keeping track of all this quite impossible.

**Comment by sgibson91** [comment]
For clarity, what I'm proposing is focussing my efforts on these todo's here: https://github.com/2i2c-org/infrastructure/issues/879#issuecomment-1024144966 Then we'll be in a position for me to open a PR that is a concrete example for us to reason about. A PR won't be coming along that will close this issue any time soon :)

**Comment by consideRatio** [comment]
Thanks @sgibson91, this issue represents a very complicated goal and I got confused!

At this point we have a clear path towards reducing the complexity by incremental progress (https://github.com/2i2c-org/infrastructure/issues/879#issuecomment-1024144966). Even though all the final details aren't planned out yet, this takes us quite far and will help us think about what to do to get us all the way.

So :100: :+1: on focusing the effort on those incremental steps at this point.

**Comment by damianavila** [comment]
Thanks @sgibson91 and @consideRatio for all the latest comments. I now have a better picture of what we were discussing 😝. I totally support going with the referenced todos so we have some ground to keep discussing, it is fun! 😉.

**Comment by sgibson91** [comment]
@2i2c-org/tech-team see below request copied from Slack!

I would really appreciate some more eyes on the CI/CD pilot Erik put together here:  https://github.com/consideRatio/testing-gh-workflows/blob/main/.github/workflows/chain_dispatcher.yaml (related file: https://github.com/consideRatio/testing-gh-workflows/blob/main/.github/workflows/chain.yaml)
• Does this feel conceptually easy to understand/maintain?
• Are we comfortable as a team implementing and maintaining this?
This will affect how I bring https://github.com/2i2c-org/infrastructure/pull/1112 into a state for review/merge I think, so I’d like some clarity from the whole team rather than keeping this conversation between myself and Erik.

**Comment by consideRatio** [comment]
IMPORTANT: @sgibson91 helped me correct a misunderstanding recently and I now don't think the strategy Sarah links to above is relevant to consider.

I think it is relevant to clarify the strategy we do aim to have detailed to the level of declaring what GitHub Workflow's we define and what jobs they include etc to accomplish our goals. For that, please go to https://github.com/2i2c-org/infrastructure/issues/1131.

**Comment by sgibson91** [comment]
Thank you for this extra thought and work @consideRatio! 🙌🏻
